### DISCUSSION TITLE
C++ Issues on Mac OS Big Sur

### DISCUSSION THREAD

**Discussion by jas220** [ImperialCollegeLondon/sharpy#147] [Q&A]
Hi, 
I've managed to successfully install SHARPy and perform the unittest. However, when I try to run a test case, I run into this error. I have already set the `write_screen` to `on`, and appended `BeamPlot` into the `flow` setting.
What do you recommend I do next?
Thanks in advance!

<img width="790" alt="Screen Shot 2021-05-10 at 1 39 48 PM" src="https://user-images.githubusercontent.com/82110485/117662077-124ac480-b197-11eb-919f-1e544e81c871.png">

**Comment by ArturoMS13** [comment]
Hi jas220,

It looks like your problem comes from the OMP library.  
Can you provide us the result of the compilation of the xbeam and UVLM libraries (the output of `make install`)?

The case you are running here is also run in the unittest so can you also double check that the unittest were successful and report the output if they were not?

Thank you

**Reply by jas220** [reply]
I've run the `cmake` and `make install` again and this is what I get now; before it looked like it had worked.
<img width="913" alt="cmake_outup" src="https://user-images.githubusercontent.com/82110485/117940635-2e6a7500-b301-11eb-8f78-3dad9dcd0dae.png">
<img width="684" alt="makeinstall_output" src="https://user-images.githubusercontent.com/82110485/117940658-33c7bf80-b301-11eb-8734-1cfbc7f5149e.png">

I have also run the unittest and this is what I get, which I thought meant it was successful.
<img width="524" alt="unittest" src="https://user-images.githubusercontent.com/82110485/117940841-65d92180-b301-11eb-90f2-c844922f9795.png">

Thank you

**Reply by ngoiz** [reply]
Before you run `cmake` make sure you clean up the build directory, otherwise you may run into conflicts... I'd suggest you clean up `build`, run `cmake`, `make install` as per the instructions. If there are any errors, please report. 

When you run the `unittest` you need to be in the correct directory (as you can see it didn't run any tests in your case because you are running the command from `bin` and there is nothing there). Quoting the documentation: 
![image](https://user-images.githubusercontent.com/35684407/117958056-eeb08700-b31a-11eb-813f-8725dba725bd.png)


If all of the above fails again, let's give it one last shot with out latest mac environment [here](https://github.com/ImperialCollegeLondon/sharpy/blob/dev_setting_error/utils/environment_macos.yml). I tried it yesterday on my mac and it worked fine (the previous one was also working, but let's give it a shot)

**Reply by jas220** [reply]
Ok, I've done everything from the beginning again and this is what I get running the command you suggested.
<img width="1232" alt="Screen Shot 2021-05-12 at 11 44 42 AM" src="https://user-images.githubusercontent.com/82110485/117964686-cd9b6680-b319-11eb-885f-67aa5c009322.png">
<img width="666" alt="Screen Shot 2021-05-12 at 11 54 51 AM" src="https://user-images.githubusercontent.com/82110485/117964710-d68c3800-b319-11eb-8a1b-743edd36a1fe.png">

Then, when I ran the unittest in the /sharpy directory, I get this.
<img width="508" alt="Screen Shot 2021-05-12 at 11 56 33 AM" src="https://user-images.githubusercontent.com/82110485/117964775-ef94e900-b319-11eb-9e02-2cc9b77fbb55.png">

**Reply by ngoiz** [reply]
Ok, the installation of the libraries has worked. The segmentation fault you are getting is likely because the version of UVLM is not the one that your current SHARPy is set up for. This would have happened because of a change in branch in either `sharpy` or `UVLM`.

So, let's try this:
1. Make sure you are running the SHARPy master branch. Run `git status` from the SHARPy directory and make sure it says you are on `master`. Else, check out the `master` branch. 
2. Run `git pull` to make sure you are at the latest version
3. Update the `UVLM` and `xbeam` submodules. Run `git submodule update --init --recursive` from the main `sharpy` directory
4. Go to `build/` and run `make install` only (i.e. no need to run `cmake ..`)
5. Run the tests again

**Reply by jas220** [reply]
Ok, I've tried following these steps and this is what I get.
<img width="519" alt="Screen Shot 2021-05-13 at 9 17 45 AM" src="https://user-images.githubusercontent.com/82110485/118099466-2ed04200-b3cd-11eb-8e33-0104119a380f.png">
<img width="572" alt="Screen Shot 2021-05-13 at 9 18 18 AM" src="https://user-images.githubusercontent.com/82110485/118099482-34c62300-b3cd-11eb-86d0-ef4b038a87af.png">
<img width="605" alt="Screen Shot 2021-05-13 at 9 18 39 AM" src="https://user-images.githubusercontent.com/82110485/118099546-49a2b680-b3cd-11eb-9efa-08594008fde9.png">
<img width="561" alt="Screen Shot 2021-05-13 at 9 22 59 AM" src="https://user-images.githubusercontent.com/82110485/118099596-5c1cf000-b3cd-11eb-9358-0c756d51afb1.png">

**Reply by ngoiz** [reply]
Ok, for some reason something has been modified in your `xbeam` installation. Now that you have all your compilers/cmake etc sorted why don't you try a clean install. Make a new folder and clone sharpy again as stated in the instructions. That way we can be sure that the folder structure is unaltered. I think that should work because you are getting the cmake process and compilation right.

If that doesn't work we'll figure something else out

**Reply by jas220** [reply]
Thank you! I will try that right now.

**Reply by jas220** [reply]
This is what I have done, and I still get the same error.
[SharpyInstallation.pdf](https://github.com/ImperialCollegeLondon/sharpy/files/6471546/SharpyInstallation.pdf)

**Reply by ngoiz** [reply]
This is strange because everything seems to have installed correctly. Thanks to @ArturoMS13 clinical eye, he spotted that you are using `zsh` in your terminal. That's the only possible difference we can identify compared to our set-up. So, final try, run `chsh -s /bin/bash` to change your terminal to `bash`. You will need to open a new terminal window in order for it to take effect (you'll see that the title of the window changes to `bash`). More info on this: https://clay-atlas.com/us/blog/2020/12/05/mac-os-en-swith-shell-script-language-zsh-bash/

Once you have done that, go to your sharpy directory and run the tests (because your installation is done properly and I think its an issue with the script running the tests).

**Reply by jas220** [reply]
I have done that, but now it cant find the sharpy_env.
<img width="567" alt="Screen Shot 2021-05-13 at 11 26 24 AM" src="https://user-images.githubusercontent.com/82110485/118113828-af4b6e80-b3de-11eb-98f8-2f0818c58f32.png">

**Reply by ngoiz** [reply]
Ok, just try installing the environment again in the bash shell. It may have been shell dependent. `conda env create -f environment_file`

**Reply by jas220** [reply]
Ok, now I get this.
<img width="1105" alt="Screen Shot 2021-05-13 at 11 48 38 AM" src="https://user-images.githubusercontent.com/82110485/118116084-ba53ce00-b3e1-11eb-84d3-59b5b4b1185a.png">

**Reply by hemayang** [reply]
Hi jas and ngoiz. Sorry for interrupting your conversations, but I am having the same issue right now. 
I really doubt if this is an issue related to the zsh/bash bcuz I was zsh before and I got the same 'segementation fault: 11' as jas's. Now I've just changed my terminal to bash but it appears the same. 
Please help and many thanks!!

**Comment by ACea15** [comment]
Dear Jesús,

I have reinstalled SHARPy in my macOS Sierra, both in bash and in the zshell, to see if I encountered the same issues. For me it went well for both shells so a few points: 

-- Changing shells is fine but make sure your Anaconda folder is ahead in the variable $PATH (by doing `echo $PATH` , otherwise do `export PATH='anaconda_folder:$PATH'` and check again). 
-- Every time you try something new go to a clean directory and start from scratch with `git clone...` to be safe, possibilities for errors grow exponentially if trying different things.
-- In the last pic you have shown you seem to have forgotten `cmake .. ` Please follow the instructions in the documentation carefully, otherwise we all are wasting our time here.
-- Before compilation (doing `cmake .. ` and so on), make sure your git working directory is clean with `git status`. If you get modified UVLM or xbeam then do `git submodule update --init --recursive` as suggested, and check again with `git status` (however if you cloned the master branch and didn't change branches this should not be the case).
-- The OpenMP warnings you are getting should not be a problem. 
-- If the problem persists you can:
      - Reach out to Imperial ICT to see if they can help with your particular problem: https://www.imperial.ac.uk/admin-services/ict/contact-ict-service-desk/
      - Install a virtual machine in your mac with a clean version of either Centos or Ubuntu and work with SHARPy from there. I guarantee that will work. 

I hope that helps and good luck!

**Reply by hemayang** [reply]
Hello Alvaro,

I am reinstalling the sharpy on my mac right now, and I just have compiled UVLM and xbeam by using `cmake` and this is what I got.
![截屏2021-05-15 16 59 55](https://user-images.githubusercontent.com/48189050/118370436-950cbe80-b59f-11eb-9b14-89c5eeec6b60.png)
As most of the responses are positive, and you said those of OpenMP should be fine, what about the highlighted `sgemm not found`? Will this be an issue? I spotted this in Jesus's screenshot as well.

Thanks.

**Reply by ACea15** [reply]
Dear Hemayang,

Don't worry about those (yet). We cannot possibly know if each output of cmake is going to be a problem. In this case I also have this warning in my mac compilation and everything is working well. Make sure the libuvlm.dylib and libxbeam.dylib are linked fine after you do `make install` and then run `python -m unittest` from ./sharpy. 
If tests do not pass, have a thorough look to all the questions and explanations posted here, and the SHARPy docs.
We have made sure compilation works in the different OS if all steps are done correctly. Thus if you keep having compilation issues after doing this, it is not a SHARPy issue but of your machine.
I refer to my previous post: a) reinstall your g++ and gfortran compilers (this is tricky so careful)  b) ask ICT for help, c) virtual machine.

Best of luck.

**Reply by jas220** [reply]
Hi Alvaro,

Thank you so much for your input, but I have a few concerns. 
-- What do you mean by "make sure Anaconda folder is ahead in the variable $PATH".  When I run `echo %PATH` I get `/Users/jesussanchez/anaconda3/bin:/Users/jesussanchez/anaconda3/condabin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin` Is that right?
-- I have reinstalled SHARPy using `git clone` every time I try something new and I still get segmentation fault: 11.
-- Sorry for the confusion, but in the last picture I was replicating the instructions given below:

> Ok, the installation of the libraries has worked. The segmentation fault you are getting is likely because the version of UVLM is not the one that your current SHARPy is set up for. This would have happened because of a change in branch in either `sharpy` or `UVLM`.
> 
> So, let's try this:
> 
> 1. Make sure you are running the SHARPy master branch. Run `git status` from the SHARPy directory and make sure it says you are on `master`. Else, check out the `master` branch.
> 2. Run `git pull` to make sure you are at the latest version
> 3. Update the `UVLM` and `xbeam` submodules. Run `git submodule update --init --recursive` from the main `sharpy` directory
> 4. Go to `build/` and run `make install` only (i.e. no need to run `cmake ..`)
> 5. Run the tests again

That is why I did not use the `cmake ..` command.  
-- I have uninstalled and reinstalled my g++ and gfortran compiler. I ran the "Hello, world!" compiler test to make sure my g++ was working. Since you've also mentioned that you use macOS Sierra, I downloaded gfortran compiler for macOS 10.2 Sierra because I was using the gfortran compiler for macOS 11 BigSurf, but I still get an error; So I'm assuming it might be some issues with macOS 11. 
-- I have also contacted Imperial ICT and they basically told me to contact you guys. However, I will contact them again so they can grant me access to install SHARPy on some computer frame that I can access from Imperial or from home. I am currently running SHARPy on an Ubuntu VM and it is very slow. Some examples that are supposed to take 30 minutes running are taking me 8+ hours, so it is not ideal, but it's better than nothing.

Once again, thank you very much for your help

**Reply by ACea15** [reply]
Jesús,

You seem to have tried everything... pitty you are having all this trouble. Can you check your git version is not very old? Just in case there could be some feature missing about submodules (very unlikely since you are getting everything but at this point...).
Your path looks ok, it can be an issue when changing shells but I don't think that's the case here. Since you reinstalled the compilers, try to reinstall Anaconda (delete your current version first) in case some dependency has been lost. You could also play with this: https://docs.conda.io/projects/conda-build/en/latest/resources/compiler-tools.html
Please google the exact error you are getting together with your macOS 11, you might get some hint of what is going on.
If all that keeps not working, maybe it helps if you speed up your VM?: https://evert.meulie.net/2017/03/26/speed-virtual-machines/
Finally, if it becomes  infeasible  for you to run some relevant cases in the VM, perhaps we can give you access to a computer so you can use SHARPy through SSH.

**Reply by jas220** [reply]
I decided to uninstall, delete, and reinstall everything (compilers and anaconda) from my mac, and start from the beginning all over again. When I run the `python -m unittest` I still get segmentation fault: 11. However, I decided to still move on to the RunningSHARPy part from the Installation Guide. Sharpy seemed to run properly.
<img width="784" alt="Screen Shot 2021-05-17 at 4 50 00 PM" src="https://user-images.githubusercontent.com/82110485/118523136-c074e200-b734-11eb-9a6b-85d6270a2c38.png">
I still get a Warning, but it keeps running. (Before I would get the same Warning but then I would get Assertion failure at kmp_csupport.cpp(590).) 
And these are the results, which coincide with the Installation guide.
<img width="570" alt="Screen Shot 2021-05-17 at 4 50 13 PM" src="https://user-images.githubusercontent.com/82110485/118523521-1c3f6b00-b735-11eb-843e-4c02093c28fd.png">
I will try to work on the Examples given as well, and see if they give the same results too.

I have still made an appointment with Imperial ICT's Mac/macOS Help Bar for tomorrow morning to try and figure this out. If not, I would really appreciate access to a computer where a can use SHARPy.

**Reply by jas220** [reply]
UPDATE: 
Although the RunningSHARPy part worked, the Examples don't. I ran them on Jupyter and they all displayed this when I would run the cell importing the required packages.
<img width="690" alt="Screen Shot 2021-05-17 at 6 12 56 PM" src="https://user-images.githubusercontent.com/82110485/118530011-100adc00-b73c-11eb-8735-881b3db9aa40.png">

**Reply by ngoiz** [reply]
Jesus, I assume this is the geradin test case. If this runs it means that your fortran installation and compilers are working properly. If you are still having trouble with the tests it is likely that the issue is in the C++ side of things (the UVLM is C++) so if you go to ICT it may be worth going over your C++ compiler installation.

To double check this, you should have no trouble running structural  only cases such as [this](https://ic-sharpy.readthedocs.io/en/master/content/example_notebooks/cantilever/static_cantilever.html) one. Then, if the compilation of the UVLM did not go well the [first test case with SHARPy](https://ic-sharpy.readthedocs.io/en/master/content/example_notebooks/simple_airfoil.html) will most likely not work.

**Reply by jas220** [reply]
Yes, the `python -m unittest` failed but the geradin test worked. Then I tried the SimpleAirfoil Example and the Kernel kept dying as soon as I ran the cell importing the different packages. Same issue with the CantiliverBeam Example.

**Reply by ACea15** [reply]
Seems guys your trouble is macOS/c++ compilers. Can you do this: 

1) delete your Anaconda distribution properly (google it and make sure you do it well)
2) Install Apple developer tools https://developer.apple.com/xcode/resources/
(it comes with different c++ compilers optimised for mac)
3) Install  the Anaconda distribution
4) Clone sharpy, create the environment...

It has to bloody work!! So do not hurry and make an understanding as you go.

**Reply by jas220** [reply]
UPDATE: I spent the last 3 hours with the Imperial ICT help desk. He concluded it might be an issue with me running on macOS BigSurf, however, he does not recommend that I go back to an older version. They asked me if SHARPy has been tested on macOS BigSurf before. We tried installing SHARPy using Docker and it worked. The `python -m unittest` run successfully. They said, however, that it was an issue having to delete the container every time I wanted to start using SHARPy again. If we tried opening the container where we had left off, SHARPy would stop working again. We needed to delete the container and relaunch the image. They tried it on a macOS Sierra and it was encountering the same issue, so they told me it might not be an issue with my mac. 
I will try to do what you just suggested; if that doesn't work, ill try to figure out how to use Docker because as of right now I don't know how to run the examples on the installation guide if I am using Docker.
The ICT help desk recommended I contact my department and ask if they could allocate for me a computer I could use remotely where I could install SHARPy on.
I will keep you posted though, thank you!

**Reply by jas220** [reply]
Hi Alvaro, 

When you say to install Apple developer tools, do you mean to install Xcode 12?

**Reply by ACea15** [reply]
Dear Jesús, did you click the link? The answer is yes, there is nothing else there but Xcode. 
We are trying to help as Good Samaritans, but our power is, unfortunately, earthbound.  
I think praying for you mac is going to be my next and final idea if a clean compilation with Xcode compilers doesn't work J.
Good luck.

**Reply by jas220** [reply]
I was just trying to make sure because I had already Xcode12 downloaded on my mac.

**Comment by ngoiz** [comment]
Maybe this is of use?

https://stackoverflow.com/a/65485591

Looks like in general there is trouble with g++ and recent OS versions. Go through the thread and give it a go (using the cxx compilers `clang++` or `gcc-10`)

**Reply by jas220** [reply]
Thank you all so much for all your help, I really appreciate it. I've tried all your recommendations but I am still getting the segmentation 11 error, but I have contacted my department to see if they could allocate a computer for me to use. For now, I will keep using a VM.
Once again, thank you very much!
